internal/controller/vrg_volrep.go

```diff
@@ -1257,31 +1258,66 @@ func (v *VRGInstance) selectVolumeReplicationClass(
 			namespacedName, err)
 	}
 
+	matchingReplicationClassList := []*volrep.VolumeReplicationClass{}
+
 	for index := range v.replClassList.Items {
 		replicationClass := &v.replClassList.Items[index]
-		if storageClass.Provisioner != replicationClass.Spec.Provisioner {
-			continue
-		}
-
 		schedulingInterval, found := replicationClass.Spec.Parameters["schedulingInterval"]
-		if !found {
-			// schedule not present in parameters of this replicationClass.
+
+		if storageClass.Provisioner != replicationClass.Spec.Provisioner || !found {
+			// skip this replication class if provisioner does not match or if schedule not found
 			continue
 		}
 
 		// ReplicationClass that matches both VRG schedule and pvc provisioner
 		if schedulingInterval == v.instance.Spec.Async.SchedulingInterval {
-			v.log.Info(fmt.Sprintf("Found VolumeReplicationClass that matches provisioner and schedule %s/%s",
-				storageClass.Provisioner, v.instance.Spec.Async.SchedulingInterval))
-
-			return replicationClass, nil
+			matchingReplicationClassList = append(matchingReplicationClassList, replicationClass)
 		}
 	}
 
-	v.log.Info(fmt.Sprintf("No VolumeReplicationClass found to match provisioner and schedule %s/%s",
+	filteredVRCs := v.filterDefaultVRC(matchingReplicationClassList)
+
+	switch len(filteredVRCs) {
+	case 0:
+		v.log.Info(fmt.Sprintf("No VolumeReplicationClass found to match provisioner and schedule %s/%s",
+			storageClass.Provisioner, v.instance.Spec.Async.SchedulingInterval))
+
+		return nil, fmt.Errorf("no VolumeReplicationClass found to match provisioner and schedule")
+	case 1:
+		v.log.Info(fmt.Sprintf("Found VolumeReplicationClass that matches provisioner and schedule %s/%s",
+			storageClass.Provisioner, v.instance.Spec.Async.SchedulingInterval))
+
+		return filteredVRCs[0], nil
+	}
+
+	v.log.Info(fmt.Sprintf("Found multiple VolumeReplicationClass that matches provisioner and schedule %s/%s",
 		storageClass.Provisioner, v.instance.Spec.Async.SchedulingInterval))
 
-	return nil, fmt.Errorf("no VolumeReplicationClass found to match provisioner and schedule")
+	return nil, fmt.Errorf("multiple VolumeReplicationClass found to match provisioner and schedule")
+}
+
+// filterDefaultVRC filters the VRC list to return VRCs with default annotation
+// if the list contains more than one VRC.
+func (v *VRGInstance) filterDefaultVRC(
+	replicationClassList []*volrep.VolumeReplicationClass,
+) []*volrep.VolumeReplicationClass {
+	if len(replicationClassList) <= 1 {
+		return replicationClassList
+	}
+
+	v.log.Info("Found multiple matching VolumeReplicationClasses, filtering with default annotation")
+
+	filteredVRCs := []*volrep.VolumeReplicationClass{}
+
+	for index := range replicationClassList {
+		if replicationClassList[index].Annotations[defaultVRCAnnotationKey] == "true" {
+			filteredVRCs = append(
+				filteredVRCs,
+				replicationClassList[index])
+		}
+	}
+
+	return filteredVRCs
```

Comment:
Here we can report the right error:
- multiple vrcs, none marked as default
- multiple vrcs, more than one marked as default

This will help to simplify the caller, and very simple and clear here - the only function dealing with multiple vrcs.

Reply:
answered in another comment above.

https://github.com/RamenDR/ramen/pull/1476#discussion_r1664046795